syncer.go

```diff
@@ -434,36 +445,39 @@ func (entry *syncerEntry) syncSecretsIndividually(names []string) []string {
 			continue
 		}
 
-		entry.writeSecret(name, secret)
+		if _, err := entry.writeSecret(name, secret); err != nil {
+			entry.Logger().WithFields(logrus.Fields{
+				"secret":   secret.Name,
+				"filename": name,
+			}).WithError(err).Error("Failed to write secret")
+		}
 	}
 	return pendingDeletions
 }
 
-func (entry *syncerEntry) writeSecret(filename string, secret *Secret) bool {
+// writeSecret writes the given secret to disk and validates it. On success, writeSecret returns true if the secret was added and false if it was changed.
+func (entry *syncerEntry) writeSecret(filename string, secret *Secret) (bool, error) {
 	state, err := entry.output.Write(secret)
 	// TODO: Filename changes of secrets might be noisy.  We should ensure they're handled more gracefully.
 	if err != nil {
-		entry.Logger().WithError(err).WithField("secret", secret.Name).Error("Failed while writing secret")
 		// This situation is unlikely: We couldn't write the secret to disk.
 		// If Output.Write fails, then no changes to the secret on-disk were made, thus we make no change
 		// to the entry.SyncState
-		return false
+		return false, err
 	}
 
 	// Success!  Store the state we wrote to disk for later validation.
 	entry.Logger().WithField("file", filename).Info("Wrote file")
+	_, present := entry.SyncState[filename]
 	entry.SyncState[filename] = *state
 
 	// Validate that we wrote our output.  This should never fail, unless there are bugs or something interfering
 	// with Keysync's output files.  It is only here to help detect problems.
 	if !entry.output.Validate(secret, *state) {
-		entry.Logger().WithField("file", filename).Error("Write succeeded, but IsValidOnDisk returned false")
-
 		// Remove inconsistent/invalid sync state, consider whatever we've written to be bad.
 		// We'll thus rewrite next iteration.
 		delete(entry.SyncState, filename)
-		return false
+		return false, fmt.Errorf("failed to validate secret %s", secret.Name)
 	}
-	// TODO: Distinguish added vs changed, using `present` above
-	return true
+	return !present, nil
```

Comment:
Just to make sure I'm understanding this right: we're relying on a file's state not being present in the `SyncState` map to indicate that it must have been changed but not added since the last sync?

Reply:
Right. The implementation is confusing and relies on some weird preconditions. Here's what I see:

- https://github.com/square/keysync/blob/master/syncer.go#L375-L383 gets the list of all secrets from keywhiz.
- For each secret, check against our `SyncState` map. If it's not in the map (see the `present` variable in the loop) or if it's `present` but changed, add it to the list of secrets to update (`needsRetrieval`).
- Fetch the secrets and contents for all of the `needsRetrieval` secrets.
- https://github.com/square/keysync/blob/master/syncer.go#L393-L395 For each of those, call this `writeSecret` helper.

So for any secret that we're writing with `writeSecret`, we know it either:
1. Wasn't in the `SyncState` and needs to be fetched or
2. It was in the `SyncState` but changed.

Reply:
Ah I see what's going on - this nested if statement is saying that we only skip adding a secret to the `needsRetrieval` list if the secret is in our `SyncState` map (i.e. it already exists in this host, when present is `true`) AND `entry.output.Validate(&secretMetadata, state)` is true (Looks like it checks if the secret "changed"): https://github.com/square/keysync/blob/master/syncer.go#L376-L383

Thanks!

Reply:
I changed the control flow to get rid of the nested `if`. Let me know if that's more readable.